### ISSUE TITLE
Video scaling (crop and fit) produces incorrect results

### ISSUE THREAD

**Issue by pkly** [theothernt/AerialViews#184]
**What were you trying to do...**
While videos play in the app, the app never scales and crops the video, instead always resizes without crop (by stretching)

**And what actually happened (ie. describe the bug)...**
It seems like the setting under playlist -> video -> video scaling always work like "scale without crop" even when set to fill with crop. 

**What kind of device do you have? (eg. Nvidia Shield, Sony TV, etc)...**
Sony Bravia Google TV xr-55a80l

unsure if important but it's all through smb

**Comment by theothernt** [comment]
Thanks for reporting the issue.

Just so I understand it properly, landscape videos are fine but with a portrait video, it's not zooming in to fill the screen but simply stretching the left and right of the video?

**Comment by theothernt** [comment]
Ah! so this was a known issue in that last version. I actually have a fix in v1.7.4 but I'm still doing some testing so it might not be out for a few weeks.

**Comment by pkly** [comment]
For me landscape videos are affected. It appears that the player resizes the video horizontally, but not vertically, thus the proportions are broken and it looks off.

I've noticed it in videos that are not quite 16:9, so they need to stretch sideways and be cropped on top and bottom, but the crop just doesn't happen, they only get resized.

**Comment by theothernt** [comment]
Btw, I have a beta on Google Play if you'd like to try it?

**Comment by pkly** [comment]
yeah so 1.7.4-beta1 still has the same issue as described above with fill and crop

**Comment by theothernt** [comment]
Ok, thanks for testing that beta.

I'll find some widescreen videos that are 4:3 and 16:10 and do some more testing.

**Comment by theothernt** [comment]
So, there is a separate VideoPlayer and ImagePlayer. The ImagePlayer uses it's own scaling/cropping. Since the last update, the VideoPlayer does too - but that doesn't mean there isn't a bug on how I handle the options, or the screen size as you suggest.

I forgot to ask before - did this bug only start in v1.7.3 ?

**Comment by pkly** [comment]
I think it was present in previous versions, but I'm not 100% sure.

**Comment by theothernt** [comment]
So I've just run a set of tests on the a) emulator and b) a Samsung Galaxy Fold virtual device (I don't have access to a TV or a Google TV device at the moment) and the videos seem to display as I'd expect.

The main examples are 1) an ultra wide screen video fits the screen and 2) an ultra tall video fits the screen - and the aspect ratio seems correct.

Can I share a link with a ZIP of these videos to test if you have time? this will help narrow down the issue to certain devices or not - I hope.

**Comment by pkly** [comment]
sure, drop it in a comment and I'll get back to you later, probably during the weekend

**Comment by theothernt** [comment]
Ok, [here is the link to the test videos](https://www.dropbox.com/scl/fi/laiibd8rgvj74ir8svsy5/test-videos.zip?rlkey=ikq1ov1gqyed80c7yawgozri4&dl=0).

All videos are freely available, with a mix of aspect ratios and frame rates.

**Comment by pkly** [comment]
Thanks for the videos, I've loaded them into the application (through smb) and the results are in. [Gallery](https://imgur.com/a/d6cXwpG)

The very tall video is glitching out on the bottom and the resize is incorrect as far as I can tell (stretched), the rest appear correct. 
For some reason you can see the Fedora logo (my pc's distro, it's currently connected, but I have absolutely no idea how it got there??), I'm guessing the size information is being read incorrectly by the video player or something?
The TV is a Sony XR-55A80L on newest software, running GoogleTV. (just had a software update few days ago too)

**Comment by theothernt** [comment]
Thanks again for all the testing. I will try and update the beta with a few more tweaks that may or may not do anything - I'll let you know when it's available.

In the meantime, the glitch where you can see the Fedora logo (from another HDMI input?) does indicate there is a bug in that version of Google TV.

My app uses Google's video player to play the video. It should take up the whole screen and even if a video doesn't fill the screen, there should be black background. Very odd!

**Comment by pkly** [comment]
I think the logo was just random garbage in video memory, as of the time when I took the picture the logo was not even visitor (was only on boot, on the bottom and yeah my PC was connected to HDMI at the time).

I'll check with the setting in a bit and let you know

**Comment by pkly** [comment]
So, here's the picture for it: https://imgur.com/a/B77weKH

It seems the issue is caused by "Enable tunneling", when I switched it off the video fit the whole screen as expected, although I'm still not sure if the player correctly scaled the video (as seen above).

I'll stay on with these settings and play more stuff that I know was broken over time. But it seems that's what was causing the issue. Fallback decoders didn't cause any changes, with tunneling enabled the video would always get screwed up at the bottom (with other random video memory this time.
I'm not exactly sure what it was but it looked like something that the application displayed much earlier, not in the same "session").

If you are simply using the built-in libraries to display the video it seems those might be bugged? I don't really need hdr and such, as none of my videos use it, but there's some higher res stuff in. I'll see if that lags later.

**Comment by theothernt** [comment]
Ok, well that's progress anyway.

I am using the Google provided player, which normally works well on most devices. The documentation leaves a lot to be desired which is why I have added options like "Enable tunnelling" to the app, although I have a rough idea what the option does - I'm not sure if all devices need it.

Given that you're using a recent device, I may change the default of that option to OFF.

Regarding the screenshot, it looks ok to me but it's hard to tell with that video. I'll find a few more very-wide and very-tall videos.

In the meantime, if the issue pops up again, let me know. And thank you so much for all your testing, there is no way I would have been able track this down otherwise!

**Comment by pkly** [comment]
Yeah no problem, glad we could find out what the issue is. If it shows up again I'll let you know.

After looking at a few more videos of my own it appears to fix the scaling issue as expected, so I think it's fine like this.

Feel free to close this issue once you change the default, or earlier if you'd like. If it shows up again in the future I'll make a new one.

**Comment by pkly** [comment]
Though I think there is one more issue - with tunneling disabled the video repeat takes like 3 second instead of happening instantly.

**Comment by theothernt** [comment]
Is this with the latest beta 4 build? I have recently changed a lot of that code, so I can go back and test it a bit more.

Also, with it only happening sometimes - could it be related to the length of the video being looped ie. longer videos take longer to start (just a guess, it shouldn't happen at the player should buffer the start of the video when looping, I think)

**Comment by pkly** [comment]
Yes, that's on beta4. It does seem related to video length and/or size (I have a few videos that are ~200mb and about 2-3 minutes long), the short videos don't seems to have issues here, so as you say it seems that the video start does not stay buffered. I've also noticed the videos stopping completely on previous versions (1.7.3 at least), where the video might pause to buffer (not for looping, just regular play)

**Comment by theothernt** [comment]
beta4 should fix that second issue you saw, there was a number of mistakes in how the 'time remaining' was calculated so it was possible to for a video to finish (ie. stop) but not skip to the next video for a while.

**Comment by theothernt** [comment]
v1.7.4 has been released on GitHub (APK), the Play Store version should be available in a week or two.
